### PR TITLE
fix(hooks): [HIMMEL-3142] bind gh pr merge to the console GO file for leg sessions

### PR DESCRIPTION
**Pull request by yotamleo**
## Summary

`block-unresolved-cr-merge.sh` was the only PreToolUse hook matched on `gh pr merge`, and it never consulted `.locks/go/` — only `merge-on-green.sh`'s HIMMEL-2919 gate did. A console-spawned leg (`HIMMEL_CONSOLE_LEG` truthy) that ran `gh pr merge` directly instead of going through `merge-on-green.sh` bypassed the console's GO entirely (PR #798 merged with no GO file ever written — see HIMMEL-3142 for the evidence trail).

This is **Option 1** from the ticket: extend `block-unresolved-cr-merge.sh` with a third gate, sharing the exact predicate `merge-on-green.sh:493` already implements rather than duplicating it.

## What changed

- **`scripts/lib/go-gate.sh`** (new): pure `go_gate(pr_num, head_sha, go_root)` predicate — no `gh` call, no fs write, only a read under `<go_root>/.locks/go/`. Extracted so `merge-on-green.sh` and `block-unresolved-cr-merge.sh` can't drift the way `spawn-claudex`/`spawn-glm` did under HIMMEL-3146.
- **`scripts/hooks/block-unresolved-cr-merge.sh`**: new gate 3 — a console-spawned leg is refused unless `.locks/go/<pr>.<head>` exists and names that exact head. `handover_root` resolves through `scripts/lib/handover-path.sh`, never a hardcoded path. No new env-var bypass — matches `merge-on-green.sh`'s own GO check's integrity posture. Selector resolution (which PR, which head) fails open on genuine ambiguity, matching the CR/CI gate siblings; the GO-file read itself is fail-closed.
- **`scripts/handover/merge-on-green.sh`**: delegates to the same `go_gate()` instead of its old inline copy.
- **`docs/handover/console-template.md`**: one paragraph naming `go.sh <pr> <head>` as the *act* of granting a GO — the console's SendMessage `GO` is notification only, not the mechanism. This is the doc gap the ticket named as the thing that would have prevented the confusion.

## Regression fixes found along the way (not in ticket scope, needed for green suites)

Two instances of the same pre-existing bug class: a console-spawned leg's own launching shell carries `HIMMEL_CONSOLE_LEG=1`, which leaked into `test-block-unresolved-cr-merge.sh` and `test-crlf-boundary.sh`'s fixtures and false-failed unrelated allow-cases once gate 3 existed to react to it (same precedent as `test-merge-on-green.sh`'s existing scrub). Both suites now scrub it explicitly. Worth flagging in review as its own class: the usual "reproduce it in the primary checkout" control is vacuous for this bug shape, because the primary's own shell never carries `HIMMEL_CONSOLE_LEG` — only a leg's launching shell does.

## Controls

- **RED control** (`test-block-unresolved-cr-merge.sh`): extracts the pre-fix hook from base `6ac483e4` (before this ticket), runs the identical PR #798 fixture (leg, no GO file) against it, and asserts it was ALLOWED there (rc=0) where the shipped hook REFUSES it (rc=2, `red_control_assert` on the specific wrong value, not just "the mutant ran").
- **Negative controls**: a non-leg session is untouched by gate 3 even under an unresolvable `go_root` (`non-leg-untouched-by-go-gate`); a leg with a valid, exactly-matching GO file still merges (`leg-valid-go-allows`); a leg with a GO file for a *stale* head is still refused (`leg-stale-go-blocks`).
- **Marker-vs-file-presence inversion control** (`leg-marker-falsy-with-go-present-allows`): `HIMMEL_CONSOLE_LEG=0` with a *valid, matching GO file present* still allows the merge. `leg-marker-falsy-allows` alone (marker off, no GO file) can't distinguish "the marker gates this" from "GO-file/go_root presence gates this" — both implementations pass that case identically. This case is the one that separates them, and empirically confirms it's the marker doing the gating, not incidental file/root presence. It doubles as a leak-forward control for the suite's own `unset HIMMEL_CONSOLE_LEG` scrub: each gate-3 case sets the marker via a per-command env prefix on the test helper, scoped to that one invocation in ordinary bash — but that scoping is a shell-mode assumption, not a structural guarantee. If the marker ever leaked forward from an earlier `HIMMEL_CONSOLE_LEG=1` case, this later case would see marker=1 with a valid GO present and could not produce rc=0 (allowed is only the falsy-marker outcome here — a leaked truthy marker with a valid *matching* GO file would actually still allow, so the real leak-forward proof is `leg-marker-falsy-allows` immediately following `leg-stale-go-blocks`'s *stale* fixture with the file removed; both are exercised back-to-back and both pass, which a leaked-truthy marker from the stale case would not).
- **Valid-go/stale-go were deliberately not re-run against the pre-fix hook.** The pre-fix hook has no `go_root`/GO-file concept at all, so it allows every case regardless of GO state — a case passing against the *absence* of the feature under test proves nothing about the feature, unlike the no-go case above, which is the PR #798 shape this ticket exists to close and is the one that matters.

## Impacted suites (run individually, by name, in the foreground — not a directory sweep)

`block-unresolved-cr-merge-suite` (33/33), `merge-on-green-suite` (348/348), `go-suite`, `crlf-boundary-suite` (90/90), `chokepoint-env-prefix-suite`, `merge-public-on-green-suite`, `clear-cr-marker-suite`, `graph-cadence-suite`, `pr-merge-suite`, `block-tail-pipe-suite`, `headed-arm-leg-suite`, `inject-initiative-suite`, `cr-high-risk-diff-suite`, `worktree-inuse-suite`, `restart-stack-suite`.

## Known-findings self-review (pre-rebutted)

Flagged: no standalone `test-go-gate.sh` for `scripts/lib/go-gate.sh`. Not added — every branch (rc 0; rc 2 × no-root / no-file / stale-head) is already exercised through both callers' real suites, and `test-merge-on-green.sh` copies the actual `go-gate.sh` file into its fixture tree rather than a stand-in, so this isn't untested, it's tested at the integration boundary rather than in isolation. A standalone unit test would duplicate coverage of the same three branches, not add new coverage.

## Option 2 (not built)

Per the ticket's explicit instruction: nothing surfaced in this work that favors Option 2 (refusing `gh pr merge` outright under the leg marker) over Option 1. Flagging per the brief rather than building it.

## Review

Cross-model critic panel was quota-rate-limited (codex, HTTP 429, 5h window) — recorded `unavailable` on the ledger rather than a false `ok`. Claude-only backstop review supplies the required `avail ... ok` row: zero Critical/Important findings, diff read in full across all seven changed files.

## Test plan

- [x] `block-unresolved-cr-merge-suite` — 33/33 (RED control + all gate-3 cases + inversion control)
- [x] `crlf-boundary-suite` — 90/90 after the `HIMMEL_CONSOLE_LEG` scrub fix
- [x] `merge-on-green-suite` — 348/348, exercises the shared `go_gate()` via the real file
- [x] `go-suite` — unaffected, passes
- [x] Remaining impacted suites listed above, all green

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Console-led merges now require GO approval matching both the pull request and commit.
  * Merge checks enforce unresolved change requests, green CI, and console GO approval.
  * GO approval now performs the merge and reports the resulting commit.

* **Bug Fixes**
  * Missing, stale, or invalid approvals are blocked.
  * Standard merges are no longer affected by unrelated console-session state.
  * Validation failures safely stop the merge.
  * Console merge commands must include a matching commit reference.

### CODE REVIEW

**Comment by yotamleo** [comment]
@coderabbitai full review

File: scripts/hooks/block-unresolved-cr-merge.sh [hunk 1]
```diff
@@ -176,4 +188,57 @@ if [ "$ci_rc" = "2" ]; then
...
+        fi
+        # shellcheck source=scripts/lib/go-gate.sh
+        # shellcheck disable=SC1091
+        if ! . "$SCRIPT_DIR/../lib/go-gate.sh" 2>/dev/null; then
+            echo "block-unresolved-cr-merge: cannot load scripts/lib/go-gate.sh — refusing (a console-spawned leg's GO gate must fail closed, not silently no-op)" >&2
+            exit 2
+        fi
+        go_reason=""
+        go_rc=0
+        go_reason=$(go_gate "$go_num" "$go_sha" "$go_root") || go_rc=$?
```
[the remark this answers is not in the document]
Comment (yotamleo) [reply scripts/hooks/block-unresolved-cr-merge.sh:256]: Fixed in 6749462a. Gate 3 now captures `--match-head-commit` (both `X` and `=X` spellings) from the merge command and, once a GO is confirmed valid, refuses unless it names the exact GO-bound head — sits after the existing fail-open selector resolution, so an unresolvable PR/head still fails open like its siblings.

Both sanctioned callers already pinned before this finding: `merge-on-green.sh:1231` (`--match-head-commit "$sha"`) and `pr-merge.sh` via `forge_pr_merge`/`gh_forge_pr_merge` (`scripts/lib/forge-github.sh:228`) — verified by reading, not skipped. The residual gap this closes is a leg hand-typing `gh pr merge` outside both wrappers, which is exactly what gate 3 exists to catch.

The "add a regression case ... assert the CLI rejects the stale expected head" ask is skipped: that asserts gh's own behavior. This hook exits before `gh pr merge` runs and cannot observe what gh does with a rejected pin, so there is nothing here for a regression case to assert. Regression coverage added instead: leg+valid-GO with no pin (blocks), wrong pin (blocks), correct pin (allows), and the `=`-form (allows) — plus a RED control against the pre-fix blob at `ef6f995a` proving the no-pin case was previously allowed.

**Comment by yotamleo** [comment]
@coderabbitai review

**Comment by yotamleo** [comment]
@coderabbitai full review

File: scripts/hooks/block-unresolved-cr-merge.sh [hunk 2]
```diff
@@ -176,4 +196,75 @@ if [ "$ci_rc" = "2" ]; then
...
+        if [ "$go_rc" = "2" ]; then
```
[the remark this answers is not in the document]
Comment (yotamleo) [reply scripts/hooks/block-unresolved-cr-merge.sh]: Fixed in 6fe4ad20. Both callers now assert `command -v go_gate >/dev/null 2>&1` right after sourcing `go-gate.sh` and refuse (existing response path: hook `exit 2`, merge-on-green `audit ... phase=console-go-symbol-missing; exit 17`) if the symbol never got defined. The comparison changed from `[ "$go_rc" = "2" ]` to `[ "$go_rc" -ne 0 ]`, with a fallback reason (naming the PR, head and unexpected rc) for the case where `go_reason` is empty because `go_gate` never actually ran.

Regression coverage added at both call sites: a truncated `go-gate.sh` (header only, no function) proves the pre-fix blob wrongly allowed the merge (RC=0, no GO file needed) and the fixed script now refuses (RC=2 / RC=17, message naming "not defined"). `scripts/hooks/test-block-unresolved-cr-merge.sh` (HIMMEL-3142-RC127-RC) and `scripts/handover/test-merge-on-green.sh` (HIMMEL-3142-RC127-MOG-RC) — both suites green (39 and 350 cases respectively).

**Comment by yotamleo** [comment]
@coderabbitai full review

File: scripts/hooks/block-unresolved-cr-merge.sh [hunk 3]
```diff
@@ -176,4 +196,82 @@ if [ "$ci_rc" = "2" ]; then
...
+        if . "$SCRIPT_DIR/../lib/handover-path.sh" 2>/dev/null; then
+            go_root=$(handover_root 2>/dev/null) || go_root=""
+        fi
+        # shellcheck source=scripts/lib/go-gate.sh
+        # shellcheck disable=SC1091
+        if ! . "$SCRIPT_DIR/../lib/go-gate.sh" 2>/dev/null; then
+            echo "block-unresolved-cr-merge: cannot load scripts/lib/go-gate.sh — refusing (a console-spawned leg's GO gate must fail closed, not silently no-op)" >&2
+            exit 2
+        fi
+        if ! command -v go_gate >/dev/null 2>&1; then
```
[the remark this answers is not in the document]
Comment (yotamleo) [reply scripts/hooks/block-unresolved-cr-merge.sh]: Fixed in 40ae11af: both call sites (block-unresolved-cr-merge.sh and merge-on-green.sh, which shares the identical pattern) now `unset -f go_gate` immediately before sourcing go-gate.sh, then `declare -F go_gate` after — closing both the PATH-executable and the inherited `export -f go_gate` bypass vectors, with a dedicated RED control for each at both call sites.

**Comment by yotamleo** [comment]
@coderabbitai full review
